cashu/core/base.py

```diff
@@ -1214,4 +1250,94 @@ def parse_obj(cls, token_dict: dict):
             u=token_dict["u"],
             t=[TokenV4Token(**t) for t in token_dict["t"]],
             d=token_dict.get("d", None),
+            r=token_dict.get("r", None),
+        )
+
+# -------- DLC STUFF --------
+
+class DiscreetLogContract(BaseModel):
+    """
+    A discrete log contract
+    """
+    settled: Optional[bool] = False
+    dlc_root: str
+    funding_amount: int
+    unit: str
+    inputs: Optional[List[Proof]] = None         # Need to verify these are indeed SCT proofs
+    debts: Optional[Dict[str, int]] = None       # We save who we owe money to here
+
+    @classmethod
+    def from_row(cls, row: Row):
+        return cls(
+            dlc_root=row["dlc_root"],
+            settled=bool(row["settled"]),
+            funding_amount=int(row["funding_amount"]),
+            unit=row["unit"],
+            debts=json.loads(row["debts"]) if row["debts"] else None,
         )
+
+class DlcBadInput(BaseModel):
+    index: int
+    detail: str
+
+class DlcFundingProof(BaseModel):
+    """
+    A dlc merkle root with its signature
+    or a dlc merkle root with bad inputs.
+    """
+    keyset: str
+    signature: str
+
+class DlcFundingAck(BaseModel):
+    dlc_root: str
+    funding_proof: DlcFundingProof
+
+class DlcFundingError(BaseModel):
+    dlc_root: str
+    bad_inputs: Optional[List[DlcBadInput]] # Used to specify potential errors
+
+class DlcOutcome(BaseModel):
+    """
+    Describes a DLC outcome
+    """
+    k: Optional[str]      # The blinded attestation secret
+    t: Optional[int]      # The timeout (claim when time is over)
+    P: str                # The payout structure associated with this outcome
+
+class DlcSettlement(BaseModel):
+    """
+    Data used to settle an outcome of a DLC
+    """
+    dlc_root: str
+    outcome: DlcOutcome
+    merkle_proof: List[str]
+
+class DlcSettlementAck(BaseModel):
+    """
+    Used by the mint to indicate the success of a DLC's funding, settlement, etc.
+    """
+    dlc_root: str
+
+class DlcSettlementError(BaseModel):
+    """
+    Indicates to the client that a DLC operation (funding, settlement, etc) failed.
+    """
+    dlc_root: str
+    details: str
+
+class DlcPayoutWitness(BaseModel):
+    # a BIP-340 signature on the root of the contract
+    signature: Optional[str] = None
+    # the discrete log of the public key (the private key)
+    secret: Optional[str] = None
+
+class DlcPayoutForm(BaseModel):
+    dlc_root: str
+    pubkey: str
+    outputs: List[BlindedMessage]
+    witness: DlcPayoutWitness
+
+class DlcPayout(BaseModel):
+    dlc_root: str
+    outputs: Optional[List[BlindedSignature]] = None
```

Comment:
This should be `signatures` not `outputs` to match the spec and be consistent with things like `PostMintResponse`